### PR TITLE
[peer review] BZ#1243740: Working with HTTP proxies

### PR DESCRIPTION
**Pull request by bfallonf**
PR in regards to https://bugzilla.redhat.com/show_bug.cgi?id=1243740

@thoraxe as this is based on info from your training docs, can I use you for some questions?
1. Is the file that the reader is modifying the buildconfig? 
2. Also, why does this differ from the procedure in the dev_guide/builds section? https://docs.openshift.org/latest/dev_guide/builds.html#using-a-proxy-for-git-cloning
   It seems to be doing the same thing... Am I wrong?
3. A lot of these are talking about both http and https. Do we have to add parameters for both? Or do you only add the parameter that's relevant to your setup?

I think that's all. 

Thanks!

### CODE REVIEW

**Comment by thoraxe** [comment]
The training docs are old and may not still be valid. I honestly don't know all the details because I've never worked with it.

That being said, there are multiple proxy "situations":

1) proxy for pods to reach outside world
2) proxy for docker daemon to reach registries
3) proxy through which git clones must occur

I'm probably munging all of those descriptions, but those are the scenarios I can think of right now. As long as we more or less cover them, we're good.

Note that that is for both authenticated as well as unauthenticated proxies, both http and https.

**Comment by thoraxe** [comment]
What you have here so far looks good. I'm just not sure it covers all the scenarios.

**Comment by bfallonf** [comment]
@thoraxe I see what you mean. There's not much I can do on my end, but I've enlisted the help of @codificat and we'll work together on getting a substantial section for the docs going. Thanks, though!

**Comment by adellape** [comment]
> we'll work together on getting a substantial section for the docs going.

@bfallonf Based on this comment, it sounds like you're planning to add more to this PR, so I'll add [WIP] to it, and you can remove/adjust when it's ready for review again? Lmk if I've misunderstood.

**Comment by codificat** [comment]
I'm sorry but I did not have time to review this to provide detailed feedback, and I'm out for the next 2 weeks so won't be able to contribute much until then.

One thing that is missing is the need to add the address of the internal registry's service to the NO_PROXY settings. At this time, by default the reference to the registry is via its service IP, and it's not possible to specify the services' whoe subnet due to RHBZ 1242961, so the IP has to be specified after creating the registry's service.

Also this section is likely to change if the RFE in BZ 1253675 gets implemented - resulting in a single place to configure proxy settings. Not available short term AFAIK but just mentioning that the doc would need updating then.

I'll try to verify the contents and/or provide additional feedback when I'm back.

**Comment by bfallonf** [comment]
@codificat Not a problem. I've sent an email to sbr-shift-list to see if anyone can help me out. It seems like a bit of a process, so I want to be able to do as much legwork as possible. I'll see what comes out of it. If there's anyone else you know that has been working with web proxies let me know.

File: admin_guide/http_proxies.adoc [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+toc::[]
+
+== Overview
+Production environments can deny direct access to the web and instead, have an HTTP or
+HTTPS proxy available. Configuring OpenShift builds and deployments to
+use these proxies is as simple as setting standard environment variables in
+configuration or JSON files.
+
+== Importing Image Streams
```
Comment (codificat) [inline admin_guide/http_proxies.adoc]: Having these env vars in the /etc/sysconfig files does not [currently] affect imagestreams, deploymentConfigs etc. So, rather than "importing image streams", the env var settings in /etc/sysconfig/* are more something like "Component configuration" but you probably can suggest a better name. It should also explain that these settings should be added to /etc/sysconfig/openshift-node (in addition to openshift-master), and maybe it's worth merging in the "Proxying Docker Pull" section below in here to keep all the settings related to /etc/sysconfig together.

The image streams / templates should be updated separately, so I'd suggest a separate section.

Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: @codificat Thanks for the tips. I see your points.

I've made the suggested changes, but kept the docker section as it's own (though closer to the above section) as from what I can tell it's still a different process for different purposes.

I received some pointers on some more info. I have to email some of the other OpenShift engineers. I'll include you on the email so that you can have some say in the process.

Thanks again.

File: admin_guide/http_proxies.adoc [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+. If you had already imported ImageStreams before creating the proxy
+configuration, delete, then re-run the process:
++
+----
+# oc delete imagestreams -n openshift --all
+# oc create -f image-streams.json -n openshift
+----
+
+== S2I Builds
+An S2I build fetches gems from a specified location. Fork that codebase and
```
Comment (bparees) [inline admin_guide/http_proxies.adoc]: gems is specific to ruby builds, i'd just say s2i builds fetch dependencies from various locations as needed.

there are also a number of ways to inject env variables into builds.  .sti/environment is one.  defining the env vars on the buildconfig itself is another which avoids having to put things in your repository.  (setting env vars in builds is discussed in the builds docs, you could just link there instead.)

File: admin_guide/http_proxies.adoc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                "env": [
+                  {
+                    "name": "HTTP_PROXY",
+                    "value": "http://USER:PASSWORD@IPADDR:PORT"
+                  },
+----
+====
+
+== Git Repository Access
+To use GitHub behind a proxy, set an environment variable on the stiStrategy:
```
Comment (bparees) [inline admin_guide/http_proxies.adoc]: this works for docker type builds too, not just s2i.

File: admin_guide/http_proxies.adoc [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        "Value": "http://USER:PASSWORD@IPADDR:PORT"
+      }
+    ]
+  }
+}
+----
+====
+
+If the variable is set on the stiStrategy it's not necessary to use the
+*_.sti/environment_* file.
```
Comment (bparees) [inline admin_guide/http_proxies.adoc]: this is wrong. this is not how you setup the proxy for git access.  please see the existing docs on how to define a proxy for git cloning, in the builds doc.

Comment (bparees) [inline admin_guide/http_proxies.adoc]: https://docs.openshift.org/latest/dev_guide/builds.html#using-a-proxy-for-git-cloning

File: admin_guide/http_proxies.adoc (hunk 2 above)
Comment (bparees) [inline admin_guide/http_proxies.adoc]: https://docs.openshift.org/latest/dev_guide/builds.html#configuring-the-source-environment

Comment (bparees) [inline admin_guide/http_proxies.adoc]: (despite how it's documented, s2i and custom builds both support having env variables added to the buildconfig).

docker type builds do not yet, though after this sprint they will.

Comment (bparees) [inline admin_guide/http_proxies.adoc]: (and only s2i builds support the .sti/environment file as one might logically assume)

Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: @bparees I put in a link to the doc you gave above, and kept some context in this doc.

File: admin_guide/http_proxies.adoc (hunk 4 above)
Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: @bparees Ah, ok. This answered one of my initial questions. Good to know it's basically a copy of the topic. I put in a copy of the section using ascii things. I feel it's best to have in both topics.

File: admin_guide/http_proxies.adoc [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+codebase and create a file named *_.sti/environment_*. The contents of the file
+are simple shell variables. Most libraries will look for `*NO_PROXY*`,
+`*HTTP_PROXY*`, and `*HTTPS_PROXY*` variables and react accordingly:
+
+S2I builds fetch dependencies from various locations as needed. The process for
+setting this up is in the link:../dev_guide/builds.html#environment-files[builds
+topic]. The contents of the file are simple shell variables. In setting an
+instance for use with web proxies, insert the following into the
+*_.sti/environment_* file. When seeing the `*NO_PROXY*`, `*HTTP_PROXY*`, and
+`*HTTPS_PROXY*` variables, OpenShift will react accordingly:
```
Comment (bparees) [inline admin_guide/http_proxies.adoc]: there's some duplication here.

what it should say:
"S2I builds fetch dependencies from various locations.  In most cases, proxies can be configured for dependency fetching by setting NO_PROXY, HTTP_PROXY, and HTTPS_PROXY environment variables.  

These environment variables can be set by......(.sti/environment file, environment config on the buildconfig definition)"

File: admin_guide/http_proxies.adoc [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+====
+
+== Git Repository Access
+
+
+
+include::dev_guide/builds.html[tag=using-a-proxy-for-git-cloning]
+
+////
+To use GitHub behind a proxy, set an environment variable on the stiStrategy:
```
Comment (bparees) [inline admin_guide/http_proxies.adoc]: on the stiStrategy or dockerStrategy.

File: admin_guide/http_proxies.adoc (hunk 4 above)
Comment (bparees) [inline admin_guide/http_proxies.adoc]: you need to remove the line about sti/environment files. (line 138-139)  it's unrelated to git proxies.

Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: @bparees Sorry, I should've mentioned that I commented it out rather than deleting it. It's handy for tracking. It's not visible in published form. What the reader will be seeing instead is a copied version of this: 

https://docs.openshift.com/enterprise/3.0/dev_guide/builds.html#using-a-proxy-for-git-cloning

File: admin_guide/http_proxies.adoc (hunk 6 above)
Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: As above, this para is commented out. 

I can delete this section actually. Doesn't really serve a purpose anymore.

Aaaand deleted.

File: admin_guide/http_proxies.adoc (hunk 5 above)
Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: Sorry, my editing process got mixed up. I usually replicate the current paragraph then make edits, then delete the older one. Just forgot to delete the older one... It's a little more edited now, with some changes.

File: admin_guide/http_proxies.adoc (hunk 4 above)
Comment (bparees) [inline admin_guide/http_proxies.adoc]: got it, looks good.

File: admin_guide/http_proxies.adoc (hunk 1 above)
Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: @codificat In looking at this as a whole, i think it's coming together. However, the "Importing Image Streams" now is empty as per your suggestion above. Is there an alternate procedure for these that I can throw into the docs? Or would someone else be a better person to ask? Thanks.

**Comment by bfallonf** [comment]
I think this is coming along. I queried @codificat inline about one point that needs fleshing out.

File: admin_guide/http_proxies.adoc [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+====
+----
+HTTP_PROXY=http://USERNAME:PASSWORD@10.0.1.1:8080/
+HTTPS_PROXY=https://USERNAME:PASSWORD@10.0.0.1:8080/
+NO_PROXY=master.hostname.example.com
+----
+====
+
+. Restart the master service:
++
```
Comment (codificat) [inline admin_guide/http_proxies.adoc]: This (restarting the master) should go in the previous section where the master is configured - add also restarting the node there.

Here we should restart the docker service instead

Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: @codificat That makes sense. I've added a step in the "hosts" section about restarting the appropriate host. 

Also, do you mean that in the "Docker" steps the master does not need to be restarted? I added that step (what it currently looks like now) after suggestion from @rimolive . Are you suggesting that they are not relevant to this section?

File: admin_guide/http_proxies.adoc [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+configuration or JSON files.
+
+[[configuring-hosts-for-proxies]]
+
+== Configuring Hosts for Proxies
+
+. Add the `*NO_PROXY*`, `*HTTP_PROXY*`, and `*HTTPS_PROXY*` environment variables
+to each host's *_/etc/sysconfig/openshift-master_* or
+*_/etc/sysconfig/openshift-node_* file depending on the type of host:
+
```
Comment (codificat) [inline admin_guide/http_proxies.adoc]: Missing a + here, breaks list numbering

File: admin_guide/http_proxies.adoc [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+[[proxying-docker-pull]]
+
+== Proxying Docker Pull
+
+Docker pulls content from the registry that hosts your built images. If you have
+an existing internal registry, use the proxy's IP with the NO_PROXY setting.
+This blacklists that registry's address, leaving the external http proxy as the
+only option.
+
```
Comment (codificat) [inline admin_guide/http_proxies.adoc]: IMHO this paragraph is unclear. Suggest rewording to: "[Docker pulls... images.]. OpenShift nodes need to perform push and pull operations to registries. If you have a Docker registry that nodes should *not* access through a proxy make sure to include that registry's address in the NO_PROXY exclusion list"

Then I would update the example below to something like this:

NO_PROXY=master.hostname.example.com,172.30.123.45,docker-registry.default.svc.cluster.local

where 172.30.123.45 is the internal docker registry's service IP, and docker-registry.default.svc.cluster.local is the service name for a service called *docker-registry* in the *default* namespace (this is the default service name for the internal registry deployed by *oadm registry*)

File: admin_guide/http_proxies.adoc [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+HTTP_PROXY=http://USERNAME:PASSWORD@10.0.1.1:8080/
+HTTPS_PROXY=https://USERNAME:PASSWORD@10.0.0.1:8080/
+NO_PROXY=master.hostname.example.com
+----
+====
+
+. Restart the master, node, and Docker services:
++
+----
+# systemctl restart openshift-master
```
Comment (codificat) [inline admin_guide/http_proxies.adoc]: No need to restart the master after changing docker settings - and I'm not sure if restarting the node is required (and if it is, I would guess that you should restart it after restarting docker)

File: admin_guide/http_proxies.adoc [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+# systemctl restart openshift-node
+# systemctl restart docker
+----
+
+. If you had already imported ImageStreams before creating the proxy
+configuration, delete, then re-run the process:
++
```
Comment (codificat) [inline admin_guide/http_proxies.adoc]: This (ImageStreams) does not belong here - I believe it's a leftover since the latest change of adding a section for default templates below. In other words: I think this should be just removed.

File: admin_guide/http_proxies.adoc [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+The following are the supported proxy environment variables with example values:
+
+====
+----
+HTTP_PROXY=http://USERNAME:PASSWORD@10.0.1.1:8080/
+HTTPS_PROXY=https://USERNAME:PASSWORD@10.0.0.1:8080/
+NO_PROXY=master.hostname.example.com
+----
+====
+
```
Comment (codificat) [inline admin_guide/http_proxies.adoc]: It's probably worth mentioning that you can also (alternatively) specify these settings in the buildConfig - this is already explained in dev_guide/builds, in "Using a Proxy for Git cloning", maybe just add a reference here?

Comment (bparees) [inline admin_guide/http_proxies.adoc]: the proxy for git cloning and the proxy for dependency downloading done by s2i are two different settings.

Comment (codificat) [inline admin_guide/http_proxies.adoc]: Ooops! sure, sorry for the confusion, was looking at two or three things at the same time and too quickly

File: admin_guide/http_proxies.adoc (hunk 9 above)
Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: @codificat I see the vagueness now that you've suggested more context. I've added the suggested info. Let me know if you want any changes.

File: admin_guide/http_proxies.adoc (hunk 11 above)
Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: Ok. Commented it out just in case.

**Comment by adellape** [comment]
@bfallonf Need anything here?

**Comment by codificat** [comment]
In addition to the current content it could be worth mentioning (in the S2I section) how to pass proxy settings to maven for maven builds: https://access.redhat.com/solutions/1758313 i.e. set a proxies section in settings.xml (either add a custom settings.xml and point to it via MAVEN_ARGS or use a custom assemble script to add it to .m2/settings.xml).

**Comment by bparees** [comment]
That does not belong in the s2i section. It belongs in the eap image doc,
given that it is not generic to s2i.

Ben Parees | OpenShift

**Comment by codificat** [comment]
Ok :)  just thought it could be worth mentioning it here because the goal of the original BZ request was to have a single place with comprehensive information about working with proxies because it's a very common deployment scenario, and eap/ews are also common. But if it makes more sense to add it elsewhere (even in the kbase where it is now) it's ok. Maybe add a link? Thanks!

**Comment by bfallonf** [comment]
@adellape Thanks for the ping. Pep and I have been talking via email and yesterday he had given me a way to test the Maven stuff he was talking about above, but I couldn't get my test machine to play nice. I was going to have a bit more of a play again today, but if we're generally agreeing that it might have a better place in the EAP section, then maybe this doc is ok as is and I can continue to work on that. 

@codificat , you're probably right in that there should be a link somewhere here. Maybe a note box saying "If you're working with EAP make sure to setup the Maven thing" with a link to the content. We can make that into a separate PR.

Unless anyone involved here has any technical edits to suggest, this might be ready for a peer review.

Thanks.

File: admin_guide/http_proxies.adoc [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+# systemctl restart openshift-node
+----
+
+[[proxying-docker-pull]]
+
+== Proxying Docker Pull
+
+OpenShift node hosts need to perform push and pull operations to Docker
+registries. If you have a registry that does not need a proxy for nodes to
+access, include the NO_PROXY parameter with the registry's hostname, the
```
Comment (tnguyen-rh) [inline admin_guide/http_proxies.adoc]: missing markup: `NO_PROXY`

File: admin_guide/http_proxies.adoc [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+NO_PROXY=master.hostname.example.com
+----
+====
+
+[[configuring-default-templates-for-proxies]]
+
+== Configuring Default Templates for Proxies
+
+The example templates available in OpenShift by default do not include settings
+for http proxies. For existing applications based on these templates, modify the
```
Comment (tnguyen-rh) [inline admin_guide/http_proxies.adoc]: upcase: HTTP

**Comment by tnguyen-rh** [comment]
@bfallonf Just a pair of small nits.  One big one, though: YAML instead of JSON -- i was under the impression that's the way things are headed.  What am i missing?

**Comment by bfallonf** [comment]
@tnguyen-rh Thanks for that. I've done edits as suggested. 

As for the YAML v JSON thing, this PR was made before we made that decision, so I didn't think of that. If it's as simple as editing the JSON bits to look like YAML this shouldn't be hard.

**Comment by tnguyen-rh** [comment]
@bfallonf 
I wrote a small tool to convert from JSON to YAML named, unsurprisingly, json2yaml.  You can find it [here](http://file.brq.redhat.com/~tnguyen/json2yaml).  If that doesn't work, i'm sure there are other options floating around the Internet...

File: admin_guide/http_proxies.adoc [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+          "manifest": {
+            "containers": [
+              {
+                "env": [
+                  {
+                    "name": "HTTP_PROXY",
+                    "value": "http://USER:PASSWORD@IPADDR:PORT"
+                  },
+----
+====
```
Comment (adellape) [inline admin_guide/http_proxies.adoc]: Couple things about this JSON. Tried converting to YAML, but the syntax was incomplete. So, I fixed it and converted to YAML, but it's v1beta1 and is therefore missing some fields. I then tried using the origin repo's origin-version-change cmd to convert it to v1, but A) I seem to recall that maybe the tool didn't really work for that old of API? and B) what I got after the conversion was not really usable.

But, I think that we can probably just show what is needed here instead of the full(er) deployment config. The point is that we're just showing what env vars need to be set, right? How about the following (the referenced stanza is based on the v1 API):

```
Set the following environment variables in a deployment configuration's `*template.spec.containers*` stanza to pass database connection information. The same can be done for configuring a pod's proxy at runtime:

====
----
containers:
- env:
  - name: "HTTP_PROXY"
    value: "http://USER:PASSWORD@IPADDR:PORT"
----
====
```

@bfallonf @tnguyen-rh WDYT?

File: admin_guide/http_proxies.adoc [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+You can also use the `oc env` command to update a DeploymentConfig with a new
+environment variable:
+
+----
+oc env <object_type>/<object_type_name> <EN_VAR>=/<VALUE>
+----
```
Comment (adellape) [inline admin_guide/http_proxies.adoc]: I feel like it may be more helpful/appropriate here to just show an actual example cmd w/ the required env var already in place, versus just showing the generic syntax, and maybe link to the dev_guide/envars topic in-line if they wanted more info on the cmd itself. Again, we're just showing what env var needs to be set, AFAICT. E.g.:

```
====
----
oc env dc/frontend HTTP_PROXY=http://USER:PASSWORD@IPADDR:PORT
----
====

File: admin_guide/http_proxies.adoc (hunk 15 above)
Comment (tnguyen-rh) [inline admin_guide/http_proxies.adoc]: @adellape I'm concerned that the less savvy reader might think this is the entirety of the data structure, copy and paste it, and run into problems.  To avoid this, i suggest adding a title to the block w/ the word "fragment" in it.  (This is what we do in the GNU project, btw.)

To really drive the point home, we could indent the thing by four spaces (two for `template`, two for `spec`) and add `[...]` above and below.  That's really noisy, however, and probably only appropriate in contexts where the example is either alone on the page, or in the midst of complete examples that might mislead the reader into thinking this one is also complete.  I don't know if this is such a context; i'm just spewing ideas.  :smile:

This whole "how and when to portray excerpts" topic is something we might chew on at the next writer's meeting...

Comment (adellape) [inline admin_guide/http_proxies.adoc]: We have other spots where a simple `...` is employed in cases like these (e.g., https://docs.openshift.com/enterprise/3.0/admin_guide/configuring_authentication.html#AllowAllPasswordIdentityProvider). I'm into it.

Re: "fragment", fine with that too. Or snippet, or something. I feel like there's another word we've maybe used elsewhere, but I'm blanking.

Comment (tnguyen-rh) [inline admin_guide/http_proxies.adoc]: @adellape Don't fight the GNU.  May the GNU be w/ you.  :smile:

Comment (adellape) [inline admin_guide/http_proxies.adoc]: :water_buffalo:

Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: That explains why I couldn't get the json to convert... I like the ideas here.  I've put in just the containers stanza (alex's above, really) and put in the ...'s. Also changed the little intro para. @tnguyen-rh @adellape thoughts?

File: admin_guide/http_proxies.adoc (hunk 16 above)
Comment (bfallonf) [inline admin_guide/http_proxies.adoc]: True. Done.

**Comment by bfallonf** [comment]
Ok. @adellape @tnguyen-rh I converted both json examples to yaml and fixed Alex's suggestion above. Let me know if there's anything else.

**Comment by adellape** [comment]
@bfallonf Did you push your changes? Also, could use a squash.

**Comment by bfallonf** [comment]
@adellape having problems with the 'include' bit. Trying to fix them now.

**Comment by bfallonf** [comment]
@adellape @tnguyen-rh dunno what happened, but something broke in the rebase. I probably went too far in. Either way, I tried to fix it and made it worse. I made a new PR ( #1057 ). Maybe give me the thumbs up or down there.

**Comment by adellape** [comment]
Closing in favor of https://github.com/openshift/openshift-docs/pull/1057
